<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Single cell RNA sequencing of vascular smooth muscle cells from atherosclerotic lesions of ApoE ko mice</title>

<script src="site_libs/header-attrs-2.29/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/yeti.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/font-awesome-6.5.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>










<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">PRJNA626450</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">
    <span class="fa fa-home"></span>
     
    Introduction
  </a>
</li>
<li>
  <a href="00.html">
    <span class="fa fa-download"></span>
     
    1. Data
  </a>
</li>
<li>
  <a href="01.html">
    <span class="fa fa-stethoscope"></span>
     
    2. QC and Preprocessing
  </a>
</li>
<li>
  <a href="02.html">
    <span class="fa fa-layer-group"></span>
     
    3. Clustering
  </a>
</li>
<li>
  <a href="03.html">
    <span class="fa fa-route"></span>
     
    4. Trajectory
  </a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Single cell RNA sequencing of vascular
smooth muscle cells from atherosclerotic lesions of ApoE ko mice</h1>

</div>


<p>This report details the comprehensive computational analysis of
single-cell RNA sequencing (scRNA-seq) data generated from
<strong>Single cell RNA sequencing of vascular smooth muscle cells from
atherosclerotic lesions of ApoE ko mice study</strong>. The primary
objective of this study was to characterize cellular heterogeneity,
identify distinct cell populations, and investigate cell state
transitions, which define the biology of the clonally expanding smooth
muscle cells (SMCs) that contribute to atherosclerotic plaques.</p>
<ul>
<li><p>The analysis pipeline commenced with raw sequencing reads
obtained from <strong>NCBI SRA:PRJNA626450</strong>. Initial
bioinformatic processing involved quantifying transcript abundances for
each cell using Salmon against a <strong>GENCODE (Release M36
(GRCm39))</strong> reference transcriptome, incorporating decoy
sequences for robust alignment and quantification. The commands used for
raw data download, reference preparation, index building, and Salmon
quantification are documented in <a href="00.html">Data Processing:
Download, Indexing, and Quantification</a>.</p></li>
<li><p>The resulting transcript-level quantifications were imported into
R <strong>using <code>tximeta</code></strong> to generate a gene-level
count matrix, represented as a <code>SingleCellExperiment</code> object.
Rigorous quality control (<a href="01.html">QC</a>) was then performed
using the <code>scater</code> package to assess library quality and
filter potentially compromised cells based on metrics such as total
library size, number of detected genes, and mitochondrial transcript
percentage, employing adaptive thresholds based on median absolute
deviations.</p></li>
<li><p>Following QC, the filtered dataset underwent standard scRNA-seq
downstream analysis steps. Counts were normalized using deconvolution
size factors (<code>scran</code>) and log-transformation
(<code>scater</code>). Feature selection identified highly variable
genes (HVGs) crucial for capturing biological variation. Dimensionality
reduction was performed using Principal Component Analysis (PCA) on the
HVGs. Subsequently, graph-based clustering (<code>scran</code>) was
applied to the principal components to partition cells into distinct
groups by (<a href="02.html">clustering</a>). These clusters were
visualized using Uniform Manifold Approximation and Projection (UMAP).
To assign biological meaning to the identified clusters, marker genes
significantly upregulated in each group were identified using
<code>scran::findMarkers</code>. The expression patterns of top markers
were visualized via heatmaps and violin plots, facilitating preliminary
annotation of cell types, such as <strong>progenitor-like, transitional,
and fibroblast-like smooth muscle cells</strong></p></li>
<li><p>Finally, to explore potential dynamic processes or developmental
relationships between the identified cell populations, <a
href="03.html">trajectory inference</a> was conducted using
<code>slingshot</code>. Based on the marker gene expression and the
hypothesis that <strong>progenitor-like cells differentiate towards a
fibroblast-like state via an intermediate transitional
population</strong>, cellular lineages were constructed, rooted in the
<strong>Progenitor-like</strong> cluster. Pseudotime values,
representing cellular progression along the inferred trajectory, were
calculated and visualized on the UMAP embedding. The expression dynamics
of key genes along pseudotime were examined to understand the molecular
changes accompanying the inferred process, and genes strongly correlated
with pseudotime were identified as potential drivers of the
transition.</p></li>
</ul>
<p>This report presents the detailed methodology, results, and
visualizations for each stage of this computational workflow, providing
insights into the cellular landscape and dynamics.</p>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
