<!DOCTYPE html>
<html lang="en-us">

<head>

  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="generator" content="Source Themes Academic 4.6.3">

  

  
  
  
  
  
    
    
    
  
  

  <meta name="author" content="Qingzhou Zhang">

  
  
  
    
  
  <meta name="description" content="This R package implements statistical modelling of affinity purification–mass spectrometry (AP-MS) data to compute confidence scores to identify bona fide protein-protein interactions (PPI).
Installation The development version can be installed through github:
devtools::install_github(repo=&quot;zqzneptune/SMAD&quot;) library(SMAD)  Input Data A demo data.frame was provided as a hint how the input data should strcutured in order to run the scoring functions:
data(TestDatInput) colnames(TestDataInput) [1] &quot;idRun&quot; &quot;idBait&quot; &quot;idPrey&quot; &quot;countPrey&quot; &quot;lenPrey&quot;     idRun idBait idPrey countPrey lenPrey     Unique ID of one AP-MS run Bait ID Prey ID Prey peptide count Protein sequence length of the prey    In case of duplcates, a suffix or prefix of e.">

  
  <link rel="alternate" hreflang="en-us" href="https://zqzneptune.github.io/post/2019-05-22-smad/">

  


  
  
  
  <meta name="theme-color" content="#3f009b">
  

  
  
  
  <script src="/js/mathjax-config.js"></script>
  

  
  
  
  
    
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/academicons/1.8.6/css/academicons.min.css" integrity="sha256-uFVgMKfistnJAfoCUQigIl+JfUaP47GrRKjf6CTPVmw=" crossorigin="anonymous">
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/all.min.css" integrity="sha256-+N4/V/SbAFiW1MPBCXnfnP9QSN3+Keu+NlB+0ev/YKQ=" crossorigin="anonymous">
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/fancybox/3.5.7/jquery.fancybox.min.css" integrity="sha256-Vzbj7sDDS/woiFS3uNKo8eIuni59rjyNGtXfstRzStA=" crossorigin="anonymous">

    
    
    
      
    
    
      
      
        
          <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.15.10/styles/github.min.css" crossorigin="anonymous" title="hl-light">
          <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.15.10/styles/dracula.min.css" crossorigin="anonymous" title="hl-dark" disabled>
        
      
    

    

    

    
    
      

      
      

      
    
      

      
      

      
    
      

      
      

      
    
      

      
      

      
    
      

      
      

      
    
      

      
      

      
    
      

      
      

      
    
      

      
      

      
    
      

      
      

      
    
      

      
      

      
    
      

      
      

      
        <script src="https://cdnjs.cloudflare.com/ajax/libs/lazysizes/5.1.2/lazysizes.min.js" integrity="sha256-Md1qLToewPeKjfAHU1zyPwOutccPAm5tahnaw7Osw0A=" crossorigin="anonymous" async></script>
      
    
      

      
      

      
    
      

      
      

      
    
      

      
      

      
        <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml.js" integrity="" crossorigin="anonymous" async></script>
      
    
      

      
      

      
    

  

  
  
  
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Cutive+Mono%7CLora:400,700%7CRoboto:400,700&display=swap">
  

  
  
  
  
  <link rel="stylesheet" href="/css/academic.css">

  





<script async src="https://www.googletagmanager.com/gtag/js?id=UA-156259873-1"></script>
<script>
  window.dataLayer = window.dataLayer || [];

  function gtag() {
      dataLayer.push(arguments);
  }

  function trackOutboundLink(url) {
    gtag('event', 'click', {
         'event_category': 'outbound',
         'event_label': url,
         'transport_type': 'beacon',
         'event_callback': function () {
           document.location = url;
         }
    });
    console.debug("Outbound link clicked: " + url);
  }

  function onClickCallback(event) {
    if ((event.target.tagName !== 'A') || (event.target.host === window.location.host)) {
      return;
    }
    trackOutboundLink(event.target);  
  }

  gtag('js', new Date());
  gtag('config', 'UA-156259873-1', {});

  
  document.addEventListener('click', onClickCallback, false);
</script>


  


  

  <link rel="manifest" href="/index.webmanifest">
  <link rel="icon" type="image/png" href="/images/icon_huf504e8a2f7c5cf9e43cec507892894a0_49847_32x32_fill_lanczos_center_2.png">
  <link rel="apple-touch-icon" type="image/png" href="/images/icon_huf504e8a2f7c5cf9e43cec507892894a0_49847_192x192_fill_lanczos_center_2.png">

  <link rel="canonical" href="https://zqzneptune.github.io/post/2019-05-22-smad/">

  
  
  
  
  
    
    
  
  
  <meta property="twitter:card" content="summary">
  
  <meta property="og:site_name" content="Predictive Systems Biology &amp; AI">
  <meta property="og:url" content="https://zqzneptune.github.io/post/2019-05-22-smad/">
  <meta property="og:title" content="Statistical Modelling of AP-MS Data (SMAD) | Predictive Systems Biology &amp; AI">
  <meta property="og:description" content="This R package implements statistical modelling of affinity purification–mass spectrometry (AP-MS) data to compute confidence scores to identify bona fide protein-protein interactions (PPI).
Installation The development version can be installed through github:
devtools::install_github(repo=&quot;zqzneptune/SMAD&quot;) library(SMAD)  Input Data A demo data.frame was provided as a hint how the input data should strcutured in order to run the scoring functions:
data(TestDatInput) colnames(TestDataInput) [1] &quot;idRun&quot; &quot;idBait&quot; &quot;idPrey&quot; &quot;countPrey&quot; &quot;lenPrey&quot;     idRun idBait idPrey countPrey lenPrey     Unique ID of one AP-MS run Bait ID Prey ID Prey peptide count Protein sequence length of the prey    In case of duplcates, a suffix or prefix of e."><meta property="og:image" content="https://zqzneptune.github.io/images/icon_huf504e8a2f7c5cf9e43cec507892894a0_49847_512x512_fill_lanczos_center_2.png">
  <meta property="twitter:image" content="https://zqzneptune.github.io/images/icon_huf504e8a2f7c5cf9e43cec507892894a0_49847_512x512_fill_lanczos_center_2.png"><meta property="og:locale" content="en-us">
  
    
      <meta property="article:published_time" content="2019-05-22T09:13:14-06:00">
    
    <meta property="article:modified_time" content="2019-05-22T09:13:14-06:00">
  

  


    






  






<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "BlogPosting",
  "mainEntityOfPage": {
    "@type": "WebPage",
    "@id": "https://zqzneptune.github.io/post/2019-05-22-smad/"
  },
  "headline": "Statistical Modelling of AP-MS Data (SMAD)",
  
  "datePublished": "2019-05-22T09:13:14-06:00",
  "dateModified": "2019-05-22T09:13:14-06:00",
  
  "author": {
    "@type": "Person",
    "name": "Qingzhou Zhang"
  },
  
  "publisher": {
    "@type": "Organization",
    "name": "Qingzhou (Johnson) Zhang | Predictive Systems Biology \u0026 AI",
    "logo": {
      "@type": "ImageObject",
      "url": "img/https://zqzneptune.github.io"
    }
  },
  "description": "This R package implements statistical modelling of affinity purification–mass spectrometry (AP-MS) data to compute confidence scores to identify bona fide protein-protein interactions (PPI).\nInstallation The development version can be installed through github:\ndevtools::install_github(repo=\u0026quot;zqzneptune/SMAD\u0026quot;) library(SMAD)  Input Data A demo data.frame was provided as a hint how the input data should strcutured in order to run the scoring functions:\ndata(TestDatInput) colnames(TestDataInput) [1] \u0026quot;idRun\u0026quot; \u0026quot;idBait\u0026quot; \u0026quot;idPrey\u0026quot; \u0026quot;countPrey\u0026quot; \u0026quot;lenPrey\u0026quot;     idRun idBait idPrey countPrey lenPrey     Unique ID of one AP-MS run Bait ID Prey ID Prey peptide count Protein sequence length of the prey    In case of duplcates, a suffix or prefix of e."
}
</script>

  

  


  


  





  <title>Statistical Modelling of AP-MS Data (SMAD) | Predictive Systems Biology &amp; AI</title>

</head>

<body id="top" data-spy="scroll" data-offset="70" data-target="#TableOfContents" >

  <aside class="search-results" id="search">
  <div class="container">
    <section class="search-header">

      <div class="row no-gutters justify-content-between mb-3">
        <div class="col-6">
          <h1>Search</h1>
        </div>
        <div class="col-6 col-search-close">
          <a class="js-search" href="#"><i class="fas fa-times-circle text-muted" aria-hidden="true"></i></a>
        </div>
      </div>

      <div id="search-box">
        
        <input name="q" id="search-query" placeholder="Search..." autocapitalize="off"
        autocomplete="off" autocorrect="off" spellcheck="false" type="search">
        
      </div>

    </section>
    <section class="section-search-results">

      <div id="search-hits">
        
      </div>

    </section>
  </div>
</aside>


  

<nav class="navbar navbar-expand-lg navbar-light compensate-for-scrollbar" id="navbar-main">
  <div class="container">

    
    
    <div class="d-none d-lg-inline-flex">
      <a class="navbar-brand" href="/">Predictive Systems Biology &amp; AI</a>
    </div>
    

    
    <button type="button" class="navbar-toggler" data-toggle="collapse"
            data-target="#navbar-content" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
    <span><i class="fas fa-bars"></i></span>
    </button>
    

    
    <div class="navbar-brand-mobile-wrapper d-inline-flex d-lg-none">
      <a class="navbar-brand" href="/">Predictive Systems Biology &amp; AI</a>
    </div>
    

    
    
    <div class="navbar-collapse main-menu-item collapse justify-content-start" id="navbar-content">

      
      <ul class="navbar-nav d-md-inline-flex">
        

        

        
        
        
          
        

        
        
        
        
        
        
          
          
          
            
          
          
        

        <li class="nav-item">
          <a class="nav-link " href="/#posts"><span>Insights</span></a>
        </li>

        
        

        

        
        
        
          
        

        
        
        
        
        
        
          
          
          
            
          
          
        

        <li class="nav-item">
          <a class="nav-link " href="/#publications"><span>Research</span></a>
        </li>

        
        

      

        
      </ul>
    </div>

    <ul class="nav-icons navbar-nav flex-row ml-auto d-flex pl-md-2">
      
      <li class="nav-item">
        <a class="nav-link js-search" href="#"><i class="fas fa-search" aria-hidden="true"></i></a>
      </li>
      

      
      <li class="nav-item">
        <a class="nav-link js-dark-toggle" href="#"><i class="fas fa-moon" aria-hidden="true"></i></a>
      </li>
      

      

    </ul>

  </div>
</nav>


  <article class="article">

  












  

  
  
  
<div class="article-container pt-3">
  <h1>Statistical Modelling of AP-MS Data (SMAD)</h1>

  

  
    


<div class="article-metadata">

  
  

  
  <span class="article-date">
    
    
      
    
    May 22, 2019
  </span>
  

  

  
  <span class="middot-divider"></span>
  <span class="article-reading-time">
    2 min read
  </span>
  

  
  
  

  
  
  <span class="middot-divider"></span>
  <span class="article-categories">
    <i class="fas fa-folder mr-1"></i><a href="/categories/software/">Software</a></span>
  

</div>

    














  
</div>



  <div class="article-container">

    <div class="article-style">
      <p>This R package implements statistical modelling of affinity purification–mass spectrometry (AP-MS) data to compute confidence scores to identify <em>bona fide</em> protein-protein interactions (PPI).</p>
<h2 id="installation">Installation</h2>
<p>The development version can be installed through github:</p>
<pre><code class="language-{r}"> devtools::install_github(repo=&quot;zqzneptune/SMAD&quot;)
 library(SMAD)
</code></pre>
<h2 id="input-data">Input Data</h2>
<p>A demo data.frame was provided as a hint how the input data should strcutured in order to run the scoring functions:</p>
<pre><code class="language-{r}">data(TestDatInput)
colnames(TestDataInput)

[1] &quot;idRun&quot; &quot;idBait&quot; &quot;idPrey&quot; &quot;countPrey&quot; &quot;lenPrey&quot; 
</code></pre>
<table>
<thead>
<tr>
<th>idRun</th>
<th align="center">idBait</th>
<th align="center">idPrey</th>
<th align="center">countPrey</th>
<th>lenPrey</th>
</tr>
</thead>
<tbody>
<tr>
<td>Unique ID of one AP-MS run</td>
<td align="center">Bait ID</td>
<td align="center">Prey ID</td>
<td align="center">Prey peptide count</td>
<td>Protein sequence length of the prey</td>
</tr>
</tbody>
</table>
<p>In case of duplcates, a suffix or prefix of e.g. &ldquo;A&rdquo;, &ldquo;B&rdquo; could be added to <strong>idRun</strong> in order to make <strong>&ldquo;idRun-idBait&rdquo;</strong> combination unique to each replicate.</p>
<h2 id="run-scoring">Run scoring</h2>
<h3 id="1-comppass">1. CompPASS</h3>
<p>Comparative Proteomic Analysis Software Suite (CompPASS) is based on spoke model. This algorithm was developed by Dr. Mathew Sowa for defining the human deubiquitinating enzyme interaction landscape <a href="https://doi.org/10.1016/j.cell.2009.04.042">(Sowa, Mathew E., et al., 2009)</a>. The implementation of this algorithm was inspired by Dr. Sowa&rsquo;s <a href="http://besra.hms.harvard.edu/ipmsmsdbs/cgi-bin/tutorial.cgi">online tutorial</a>. The output includes Z-score, S-score, D-score and WD-score. In its implementation in BioPlex 1.0 <a href="https://doi.org/10.1016/j.cell.2015.06.043">(Huttlin, Edward L., et al., 2015)</a> and
BioPlex 2.0 <a href="https://www.nature.com/articles/nature22366">(Huttlin, Edward L., et al., 2017)</a>, a naive
Bayes classifier that learns to distinguish true interacting proteins from
non-specific background and false positive identifications was included in the
compPASS pipline. This function was optimized from the <a href="https://github.com/dnusinow/cRomppass">source code</a>.</p>
<p>The input data.frame, <em>datInput</em>, should include:<strong>idRun</strong>, <strong>idBait</strong>, <strong>idPrey</strong> and <strong>countPrey</strong>.</p>
<pre><code class="language-{r}">datScore &lt;- CompPASS(datInput)
</code></pre>
<h3 id="2-dice">2. DICE</h3>
<p>The Dice coefficient is used to score the interaction scores across prey pair-wise combinations, which was proposed by <a href="https://doi.org/10.1093/bioinformatics/btn036">(Bing Zhang et al., 2008)</a></p>
<p>The input data.frame, <em>datInput</em>, should include:<strong>idRun</strong> and <strong>idPrey</strong>.</p>
<pre><code class="language-{r}">datScore &lt;- DICE(datInput)
</code></pre>
<h3 id="3-hart">3. Hart</h3>
<p>Hart scoring algorithm is based on a hypergeometric distribution error model <a href="https://doi.org/10.1186/1471-2105-8-236">(Hart et al., 2007)</a>.</p>
<p>The input data.frame, <em>datInput</em>, should include:<strong>idRun</strong> and <strong>idPrey</strong>.</p>
<pre><code class="language-{r}">datScore &lt;- Hart(datInput)
</code></pre>
<h3 id="4-hgscore">4. HGScore</h3>
<p>HGScore algorithm is based on a hypergeometric distribution error model <a href="https://doi.org/10.1186/1471-2105-8-236">(Hart et al., 2007)</a> with incorporation of NSAF <a href="https://doi.org/10.1021/pr060161n">(Zybailov, Boris, et al., 2006)</a>. This algorithm was first introduced to predict the protein complex network of Drosophila melanogaster <a href="https://doi.org/10.1016/j.cell.2011.08.047">(Guruharsha, K. G., et al., 2011)</a>. This scoring algorithm was based on matrix model.</p>
<p>The input data.frame, <em>datInput</em>, should include:<strong>idRun</strong>, <strong>idPrey</strong>, <strong>countPrey</strong> and <strong>lenPrey</strong>.</p>
<pre><code class="language-{r}">datScore &lt;- HG(datInput)
</code></pre>
<h3 id="5-pe">5. PE</h3>
<p>PE incorporated both spoke and matrix model as repored in <a href="https://doi.org/10.1074/mcp.M600381-MCP200">(Sean R. Collins, et al., 2007)</a>.</p>
<p>The input data.frame, <em>datInput</em>, should include:<strong>idRun</strong>, <strong>idBait</strong> and <strong>idPrey</strong>.</p>
<pre><code class="language-{r}">datScore &lt;- PE(datInput)
</code></pre>
<h2 id="license">License</h2>
<p>MIT @ Qingzhou Zhang</p>

    </div>

    



<div class="article-tags">
  
  <a class="badge badge-light" href="/tags/r/">R</a>
  
  <a class="badge badge-light" href="/tags/proteomics/">Proteomics</a>
  
  <a class="badge badge-light" href="/tags/interactome/">Interactome</a>
  
  <a class="badge badge-light" href="/tags/ap-ms/">AP-MS</a>
  
</div>



<div class="share-box" aria-hidden="true">
  <ul class="share">
    
      
      
      
        
      
      
      
      <li>
        <a href="https://twitter.com/intent/tweet?url=https://zqzneptune.github.io/post/2019-05-22-smad/&amp;text=Statistical%20Modelling%20of%20AP-MS%20Data%20%28SMAD%29" target="_blank" rel="noopener" class="share-btn-twitter">
          <i class="fab fa-twitter"></i>
        </a>
      </li>
    
      
      
      
        
      
      
      
      <li>
        <a href="https://www.facebook.com/sharer.php?u=https://zqzneptune.github.io/post/2019-05-22-smad/&amp;t=Statistical%20Modelling%20of%20AP-MS%20Data%20%28SMAD%29" target="_blank" rel="noopener" class="share-btn-facebook">
          <i class="fab fa-facebook"></i>
        </a>
      </li>
    
      
      
      
        
      
      
      
      <li>
        <a href="mailto:?subject=Statistical%20Modelling%20of%20AP-MS%20Data%20%28SMAD%29&amp;body=https://zqzneptune.github.io/post/2019-05-22-smad/" target="_blank" rel="noopener" class="share-btn-email">
          <i class="fas fa-envelope"></i>
        </a>
      </li>
    
      
      
      
        
      
      
      
      <li>
        <a href="https://www.linkedin.com/shareArticle?url=https://zqzneptune.github.io/post/2019-05-22-smad/&amp;title=Statistical%20Modelling%20of%20AP-MS%20Data%20%28SMAD%29" target="_blank" rel="noopener" class="share-btn-linkedin">
          <i class="fab fa-linkedin-in"></i>
        </a>
      </li>
    
      
      
      
        
      
      
      
      <li>
        <a href="https://web.whatsapp.com/send?text=Statistical%20Modelling%20of%20AP-MS%20Data%20%28SMAD%29%20https://zqzneptune.github.io/post/2019-05-22-smad/" target="_blank" rel="noopener" class="share-btn-whatsapp">
          <i class="fab fa-whatsapp"></i>
        </a>
      </li>
    
      
      
      
        
      
      
      
      <li>
        <a href="https://service.weibo.com/share/share.php?url=https://zqzneptune.github.io/post/2019-05-22-smad/&amp;title=Statistical%20Modelling%20of%20AP-MS%20Data%20%28SMAD%29" target="_blank" rel="noopener" class="share-btn-weibo">
          <i class="fab fa-weibo"></i>
        </a>
      </li>
    
  </ul>
</div>












  






  
  
  
    
  
  
  <div class="media author-card content-widget-hr">
    
      
      <img class="portrait mr-3" src="/authors/admin/avatar_hub8d53297f87883d816c4cfd9dec236f6_774202_250x250_fill_q90_lanczos_center.jpg" alt="Avatar">
    

    <div class="media-body">
      <h5 class="card-title"><a href="https://zqzneptune.github.io">Qingzhou Zhang</a></h5>
      <h6 class="card-subtitle">A computational systems biologist</h6>
      
      <ul class="network-icon" aria-hidden="true">
  
    
    
    
      
    
    
    
    
    
      
    
    <li>
      <a href="https://github.com/zqzneptune" target="_blank" rel="noopener">
        <i class="fab fa-github"></i>
      </a>
    </li>
  
</ul>

    </div>
  </div>









  
  
  <div class="article-widget content-widget-hr">
    <h3>Related</h3>
    <ul>
      
      <li><a href="/post/2018-07-08-apms/">Awsome affinity purification mass spectrometry (AP-MS)</a></li>
      
      <li><a href="/publication/nbt.4024/">Global landscape of cell envelope protein complexes in Escherichia coli</a></li>
      
      <li><a href="/publication/j.cels.2017.10.010/">A Map of Human Mitochondrial Protein Interactions Linked to Neurodegeneration Reveals New Mechanisms of Redox Homeostasis and NF-κB Signaling</a></li>
      
      <li><a href="/publication/oncotarget.22413/">Renal oncocytoma characterized by the defective complex I of the respiratory chain boosts the synthesis of the ROS scavenger glutathione</a></li>
      
      <li><a href="/post/2014-09-23-rged/">Hello, RGED!</a></li>
      
    </ul>
  </div>
  



  </div>
</article>

      

    
    
    
      <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery.imagesloaded/4.1.4/imagesloaded.pkgd.min.js" integrity="sha256-lqvxZrPLtfffUl2G/e7szqSvPBILGbwmsGE1MKlOi0Q=" crossorigin="anonymous"></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery.isotope/3.0.6/isotope.pkgd.min.js" integrity="sha256-CBrpuqrMhXwcLLUd5tvQ4euBHCdh7wGlDfNz8vbu/iI=" crossorigin="anonymous"></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/fancybox/3.5.7/jquery.fancybox.min.js" integrity="sha256-yt2kYMy0w8AbtF89WXb2P1rfjcP/HTHLT7097U8Y5b8=" crossorigin="anonymous"></script>

      

      
        
        <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.15.10/highlight.min.js" integrity="sha256-1zu+3BnLYV9LdiY85uXMzii3bdrkelyp37e0ZyTAQh0=" crossorigin="anonymous"></script>
        
        <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.15.10/languages/r.min.js"></script>
        
      

    

    
    

    
    
    <script>const code_highlighting = true;</script>
    

    
    
    
    
    
    
    <script>
      const search_config = {"indexURI":"/index.json","minLength":1,"threshold":0.3};
      const i18n = {"no_results":"No results found","placeholder":"Search...","results":"results found"};
      const content_type = {
        'post': "Posts",
        'project': "Projects",
        'publication' : "Publications",
        'talk' : "Talks"
        };
    </script>
    

    
    

    
    
    <script id="search-hit-fuse-template" type="text/x-template">
      <div class="search-hit" id="summary-{{key}}">
      <div class="search-hit-content">
        <div class="search-hit-name">
          <a href="{{relpermalink}}">{{title}}</a>
          <div class="article-metadata search-hit-type">{{type}}</div>
          <p class="search-hit-description">{{snippet}}</p>
        </div>
      </div>
      </div>
    </script>
    

    
    
    <script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/3.2.1/fuse.min.js" integrity="sha256-VzgmKYmhsGNNN4Ph1kMW+BjoYJM2jV5i4IlFoeZA9XI=" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script>
    

    
    

    
    

    
    
    
    
    
    
    
    
    
      
    
    
    
    
    <script src="/js/academic.min.600a2e440c16093e23e26e90951c4d4b.js"></script>

    






  
  
  <div class="container">
    <footer class="site-footer">
  

  <p class="powered-by">
    ©2014-2025 Qingzhou Zhang (CC) BY-NC-SA 4.0. All thoughts and opinions here are my own. &middot; 

    Powered by <a href="https://gohugo.io" target="_blank" rel="noopener">Hugo</a>.

    
    <span class="float-right" aria-hidden="true">
      <a href="#" class="back-to-top">
        <span class="button_icon">
          <i class="fas fa-chevron-up fa-2x"></i>
        </span>
      </a>
    </span>
    
  </p>
</footer>

  </div>
  

  
<div id="modal" class="modal fade" role="dialog">
  <div class="modal-dialog">
    <div class="modal-content">
      <div class="modal-header">
        <h5 class="modal-title">Cite</h5>
        <button type="button" class="close" data-dismiss="modal" aria-label="Close">
          <span aria-hidden="true">&times;</span>
        </button>
      </div>
      <div class="modal-body">
        <pre><code class="tex hljs"></code></pre>
      </div>
      <div class="modal-footer">
        <a class="btn btn-outline-primary my-1 js-copy-cite" href="#" target="_blank">
          <i class="fas fa-copy"></i> Copy
        </a>
        <a class="btn btn-outline-primary my-1 js-download-cite" href="#" target="_blank">
          <i class="fas fa-download"></i> Download
        </a>
        <div id="modal-error"></div>
      </div>
    </div>
  </div>
</div>

</body>
</html>
