<!DOCTYPE html>
<html lang="en-us">

<head>

  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="generator" content="Source Themes Academic 4.6.3">

  

  
  
  
  
  
    
    
    
  
  

  <meta name="author" content="Q. (Johnson) Zhang">

  
  
  
    
  
  <meta name="description" content="The SMAD (Statistical Modelling of AP-MS Data) aims to assgin scores to protein-protein interactions. Currently, there are two algorithms implemented as in seperated functions: CompPASS and HG.
Installation Use Bioconductor:
BiocManager::install(&quot;SMAD&quot;)  Alternatively, you can install from my Github repository:
devtools::install_github(&quot;zqzneptune/SMAD&quot;)  Quick start Loading the package:
library(SMAD)  Use the test data delivered with the package:
data(&quot;TestDatInput&quot;) head(TestDatInput) ## idRun idBait idPrey countPrey lenPrey ## 1 70380 ISG20 ARPC2 17 300 ## 2 70380 ISG20 RPL4 8 427 ## 3 70380 ISG20 MARCKSL1 4 195 ## 4 70380 ISG20 RCN1 4 331 ## 5 70380 ISG20 YBX1 9 324 ## 6 70380 ISG20 BASP1 6 227  The test data is for demo only, there are in total 65 AP-MS expriments covering 65 bait proteins and 26441 bait-prey pairs.">

  
  <link rel="alternate" hreflang="en-us" href="https://zqzneptune.github.io/post/2019-05-22-smad/">

  


  
  
  
  <meta name="theme-color" content="#3f009b">
  

  
  
  
  <script src="/js/mathjax-config.js"></script>
  

  
  
  
  
    
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/academicons/1.8.6/css/academicons.min.css" integrity="sha256-uFVgMKfistnJAfoCUQigIl+JfUaP47GrRKjf6CTPVmw=" crossorigin="anonymous">
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/all.min.css" integrity="sha256-+N4/V/SbAFiW1MPBCXnfnP9QSN3+Keu+NlB+0ev/YKQ=" crossorigin="anonymous">
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/fancybox/3.5.7/jquery.fancybox.min.css" integrity="sha256-Vzbj7sDDS/woiFS3uNKo8eIuni59rjyNGtXfstRzStA=" crossorigin="anonymous">

    
    
    
      
    
    
      
      
        
          <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.15.10/styles/github.min.css" crossorigin="anonymous" title="hl-light">
          <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.15.10/styles/dracula.min.css" crossorigin="anonymous" title="hl-dark" disabled>
        
      
    

    

    

    
    
      

      
      

      
    
      

      
      

      
    
      

      
      

      
    
      

      
      

      
    
      

      
      

      
    
      

      
      

      
    
      

      
      

      
    
      

      
      

      
    
      

      
      

      
    
      

      
      

      
    
      

      
      

      
        <script src="https://cdnjs.cloudflare.com/ajax/libs/lazysizes/5.1.2/lazysizes.min.js" integrity="sha256-Md1qLToewPeKjfAHU1zyPwOutccPAm5tahnaw7Osw0A=" crossorigin="anonymous" async></script>
      
    
      

      
      

      
    
      

      
      

      
    
      

      
      

      
        <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml.js" integrity="" crossorigin="anonymous" async></script>
      
    
      

      
      

      
    

  

  
  
  
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Cutive+Mono%7CLora:400,700%7CRoboto:400,700&display=swap">
  

  
  
  
  
  <link rel="stylesheet" href="/css/academic.css">

  





<script async src="https://www.googletagmanager.com/gtag/js?id=UA-156259873-1"></script>
<script>
  window.dataLayer = window.dataLayer || [];

  function gtag() {
      dataLayer.push(arguments);
  }

  function trackOutboundLink(url) {
    gtag('event', 'click', {
         'event_category': 'outbound',
         'event_label': url,
         'transport_type': 'beacon',
         'event_callback': function () {
           document.location = url;
         }
    });
    console.debug("Outbound link clicked: " + url);
  }

  function onClickCallback(event) {
    if ((event.target.tagName !== 'A') || (event.target.host === window.location.host)) {
      return;
    }
    trackOutboundLink(event.target);  
  }

  gtag('js', new Date());
  gtag('config', 'UA-156259873-1', {});

  
  document.addEventListener('click', onClickCallback, false);
</script>


  


  

  <link rel="manifest" href="/index.webmanifest">
  <link rel="icon" type="image/png" href="/images/icon_huf504e8a2f7c5cf9e43cec507892894a0_49847_32x32_fill_lanczos_center_2.png">
  <link rel="apple-touch-icon" type="image/png" href="/images/icon_huf504e8a2f7c5cf9e43cec507892894a0_49847_192x192_fill_lanczos_center_2.png">

  <link rel="canonical" href="https://zqzneptune.github.io/post/2019-05-22-smad/">

  
  
  
  
  
    
    
  
  
  <meta property="twitter:card" content="summary">
  
  <meta property="og:site_name" content="zqzneptune">
  <meta property="og:url" content="https://zqzneptune.github.io/post/2019-05-22-smad/">
  <meta property="og:title" content="SMAD | zqzneptune">
  <meta property="og:description" content="The SMAD (Statistical Modelling of AP-MS Data) aims to assgin scores to protein-protein interactions. Currently, there are two algorithms implemented as in seperated functions: CompPASS and HG.
Installation Use Bioconductor:
BiocManager::install(&quot;SMAD&quot;)  Alternatively, you can install from my Github repository:
devtools::install_github(&quot;zqzneptune/SMAD&quot;)  Quick start Loading the package:
library(SMAD)  Use the test data delivered with the package:
data(&quot;TestDatInput&quot;) head(TestDatInput) ## idRun idBait idPrey countPrey lenPrey ## 1 70380 ISG20 ARPC2 17 300 ## 2 70380 ISG20 RPL4 8 427 ## 3 70380 ISG20 MARCKSL1 4 195 ## 4 70380 ISG20 RCN1 4 331 ## 5 70380 ISG20 YBX1 9 324 ## 6 70380 ISG20 BASP1 6 227  The test data is for demo only, there are in total 65 AP-MS expriments covering 65 bait proteins and 26441 bait-prey pairs."><meta property="og:image" content="https://zqzneptune.github.io/images/icon_huf504e8a2f7c5cf9e43cec507892894a0_49847_512x512_fill_lanczos_center_2.png">
  <meta property="twitter:image" content="https://zqzneptune.github.io/images/icon_huf504e8a2f7c5cf9e43cec507892894a0_49847_512x512_fill_lanczos_center_2.png"><meta property="og:locale" content="en-us">
  
    
      <meta property="article:published_time" content="2019-05-22T09:13:14-06:00">
    
    <meta property="article:modified_time" content="2019-05-22T09:13:14-06:00">
  

  


    






  






<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "BlogPosting",
  "mainEntityOfPage": {
    "@type": "WebPage",
    "@id": "https://zqzneptune.github.io/post/2019-05-22-smad/"
  },
  "headline": "SMAD",
  
  "datePublished": "2019-05-22T09:13:14-06:00",
  "dateModified": "2019-05-22T09:13:14-06:00",
  
  "author": {
    "@type": "Person",
    "name": "Q. (Johnson) Zhang"
  },
  
  "publisher": {
    "@type": "Organization",
    "name": "zqzneptune",
    "logo": {
      "@type": "ImageObject",
      "url": "img/https://zqzneptune.github.io"
    }
  },
  "description": "The SMAD (Statistical Modelling of AP-MS Data) aims to assgin scores to protein-protein interactions. Currently, there are two algorithms implemented as in seperated functions: CompPASS and HG.\nInstallation Use Bioconductor:\nBiocManager::install(\u0026quot;SMAD\u0026quot;)  Alternatively, you can install from my Github repository:\ndevtools::install_github(\u0026quot;zqzneptune/SMAD\u0026quot;)  Quick start Loading the package:\nlibrary(SMAD)  Use the test data delivered with the package:\ndata(\u0026quot;TestDatInput\u0026quot;) head(TestDatInput) ## idRun idBait idPrey countPrey lenPrey ## 1 70380 ISG20 ARPC2 17 300 ## 2 70380 ISG20 RPL4 8 427 ## 3 70380 ISG20 MARCKSL1 4 195 ## 4 70380 ISG20 RCN1 4 331 ## 5 70380 ISG20 YBX1 9 324 ## 6 70380 ISG20 BASP1 6 227  The test data is for demo only, there are in total 65 AP-MS expriments covering 65 bait proteins and 26441 bait-prey pairs."
}
</script>

  

  


  


  





  <title>SMAD | zqzneptune</title>

</head>

<body id="top" data-spy="scroll" data-offset="70" data-target="#TableOfContents" >

  <aside class="search-results" id="search">
  <div class="container">
    <section class="search-header">

      <div class="row no-gutters justify-content-between mb-3">
        <div class="col-6">
          <h1>Search</h1>
        </div>
        <div class="col-6 col-search-close">
          <a class="js-search" href="#"><i class="fas fa-times-circle text-muted" aria-hidden="true"></i></a>
        </div>
      </div>

      <div id="search-box">
        
        <input name="q" id="search-query" placeholder="Search..." autocapitalize="off"
        autocomplete="off" autocorrect="off" spellcheck="false" type="search">
        
      </div>

    </section>
    <section class="section-search-results">

      <div id="search-hits">
        
      </div>

    </section>
  </div>
</aside>


  

<nav class="navbar navbar-expand-lg navbar-light compensate-for-scrollbar" id="navbar-main">
  <div class="container">

    
    
    <div class="d-none d-lg-inline-flex">
      <a class="navbar-brand" href="/">zqzneptune</a>
    </div>
    

    

    
    <div class="navbar-brand-mobile-wrapper d-inline-flex d-lg-none">
      <a class="navbar-brand" href="/">zqzneptune</a>
    </div>
    

    
    
    <div class="navbar-collapse main-menu-item collapse justify-content-start" id="navbar-content">

      
      <ul class="navbar-nav d-md-inline-flex">
        

      

        
      </ul>
    </div>

    <ul class="nav-icons navbar-nav flex-row ml-auto d-flex pl-md-2">
      
      <li class="nav-item">
        <a class="nav-link js-search" href="#"><i class="fas fa-search" aria-hidden="true"></i></a>
      </li>
      

      
      <li class="nav-item">
        <a class="nav-link js-dark-toggle" href="#"><i class="fas fa-moon" aria-hidden="true"></i></a>
      </li>
      

      

    </ul>

  </div>
</nav>


  <article class="article">

  












  

  
  
  
<div class="article-container pt-3">
  <h1>SMAD</h1>

  

  
    


<div class="article-metadata">

  
  

  
  <span class="article-date">
    
    
      
    
    May 22, 2019
  </span>
  

  

  
  <span class="middot-divider"></span>
  <span class="article-reading-time">
    2 min read
  </span>
  

  
  
  

  
  
  <span class="middot-divider"></span>
  <span class="article-categories">
    <i class="fas fa-folder mr-1"></i><a href="/categories/software/">Software</a></span>
  

</div>

    














  
</div>



  <div class="article-container">

    <div class="article-style">
      <p>The SMAD (Statistical Modelling of AP-MS Data) aims to assgin scores to
protein-protein interactions. Currently, there are two algorithms
implemented as in seperated functions: <strong>CompPASS</strong> and <strong>HG</strong>.</p>
<h2 id="installation">Installation</h2>
<p>Use Bioconductor:</p>
<pre><code>BiocManager::install(&quot;SMAD&quot;)
</code></pre>
<p>Alternatively, you can install from my Github repository:</p>
<pre><code>devtools::install_github(&quot;zqzneptune/SMAD&quot;)
</code></pre>
<h2 id="quick-start">Quick start</h2>
<p>Loading the package:</p>
<pre><code>library(SMAD)
</code></pre>
<p>Use the test data delivered with the package:</p>
<pre><code>data(&quot;TestDatInput&quot;)
head(TestDatInput)

##   idRun idBait   idPrey countPrey lenPrey
## 1 70380  ISG20    ARPC2        17     300
## 2 70380  ISG20     RPL4         8     427
## 3 70380  ISG20 MARCKSL1         4     195
## 4 70380  ISG20     RCN1         4     331
## 5 70380  ISG20     YBX1         9     324
## 6 70380  ISG20    BASP1         6     227
</code></pre>
<p>The test data is for demo only, there are in total <strong>65</strong> AP-MS
expriments covering <strong>65</strong> bait proteins and <strong>26441</strong> bait-prey pairs.</p>
<h2 id="comppass">CompPASS</h2>
<p>Comparative Proteomic Analysis Software Suite (CompPASS) is based on
spoke model. The implementation of this algorithm was inspired by
<a href="http://besra.hms.harvard.edu/ipmsmsdbs/cgi-bin/tutorial.cgi">Dr. Sowa’s online
tutorial</a>.
The output includes Z-score, S-score, D-score and WD-score.</p>
<pre><code>scoreCompPASS &lt;- CompPASS(TestDatInput)
head(scoreCompPASS)

##   idBait idPrey AvePSM      scoreZ    scoreS    scoreD Entropy    scoreWD
## 1  AIFM3   ACTB     19 -0.62875216  4.358899  4.358899       0 0.05209068
## 2  AIFM3  ACTC1     15  0.03766105  4.135851  4.135851       0 0.06942625
## 3  AIFM3  ACTN2      2 -0.36323536  2.081666  2.081666       0 0.05954275
## 4  AIFM3  ACTN4      5 -1.00689296  2.271284  2.271284       0 0.05000186
## 5  AIFM3   AHCY      6 -0.58996199  2.571025  2.571025       0 0.04553311
## 6  AIFM3  AIFM1     20  1.95142442 10.871146 10.871146       0 0.41566770
</code></pre>
<h2 id="hgscore">HGscore</h2>
<p>HGScore scoring algorithm used a hypergeometric distribution error model
with incorporation of NSAF. As a matrix model, it computes prey-prey
associations. Unlike CompPASS, we need protein length for each prey in
the additional column.</p>
<pre><code>scoreHG &lt;- HG(TestDatInput)
head(scoreHG)

##   InteractorA InteractorB         HG
## 1         A2M       ABCB1 8.27071788
## 2         A2M       ABCC4 7.18157099
## 3         A2M       ABCD3 0.75126610
## 4         A2M      ACADSB 2.70799149
## 5         A2M       ACAT1 0.09892875
## 6         A2M        ACLY 0.38878476
</code></pre>
<h2 id="conclusion">Conclusion</h2>
<p>SMAD is actively under development, and will include more scoring
algorithms. Stay tuned for updates.</p>

    </div>

    



<div class="article-tags">
  
  <a class="badge badge-light" href="/tags/r/">R</a>
  
  <a class="badge badge-light" href="/tags/proteomics/">Proteomics</a>
  
  <a class="badge badge-light" href="/tags/interactome/">Interactome</a>
  
  <a class="badge badge-light" href="/tags/ap-ms/">AP-MS</a>
  
</div>



<div class="share-box" aria-hidden="true">
  <ul class="share">
    
      
      
      
        
      
      
      
      <li>
        <a href="https://twitter.com/intent/tweet?url=https://zqzneptune.github.io/post/2019-05-22-smad/&amp;text=SMAD" target="_blank" rel="noopener" class="share-btn-twitter">
          <i class="fab fa-twitter"></i>
        </a>
      </li>
    
      
      
      
        
      
      
      
      <li>
        <a href="https://www.facebook.com/sharer.php?u=https://zqzneptune.github.io/post/2019-05-22-smad/&amp;t=SMAD" target="_blank" rel="noopener" class="share-btn-facebook">
          <i class="fab fa-facebook"></i>
        </a>
      </li>
    
      
      
      
        
      
      
      
      <li>
        <a href="mailto:?subject=SMAD&amp;body=https://zqzneptune.github.io/post/2019-05-22-smad/" target="_blank" rel="noopener" class="share-btn-email">
          <i class="fas fa-envelope"></i>
        </a>
      </li>
    
      
      
      
        
      
      
      
      <li>
        <a href="https://www.linkedin.com/shareArticle?url=https://zqzneptune.github.io/post/2019-05-22-smad/&amp;title=SMAD" target="_blank" rel="noopener" class="share-btn-linkedin">
          <i class="fab fa-linkedin-in"></i>
        </a>
      </li>
    
      
      
      
        
      
      
      
      <li>
        <a href="https://web.whatsapp.com/send?text=SMAD%20https://zqzneptune.github.io/post/2019-05-22-smad/" target="_blank" rel="noopener" class="share-btn-whatsapp">
          <i class="fab fa-whatsapp"></i>
        </a>
      </li>
    
      
      
      
        
      
      
      
      <li>
        <a href="https://service.weibo.com/share/share.php?url=https://zqzneptune.github.io/post/2019-05-22-smad/&amp;title=SMAD" target="_blank" rel="noopener" class="share-btn-weibo">
          <i class="fab fa-weibo"></i>
        </a>
      </li>
    
  </ul>
</div>












  






  
  
  
    
  
  
  <div class="media author-card content-widget-hr">
    
      
      <img class="portrait mr-3" src="/authors/admin/avatar_hu8cbbf145fba7419415e10c5e5a434d3a_30676_250x250_fill_q90_lanczos_center.jpg" alt="Avatar">
    

    <div class="media-body">
      <h5 class="card-title"><a href="https://zqzneptune.github.io">Q. (Johnson) Zhang</a></h5>
      <h6 class="card-subtitle">A computational biologist</h6>
      <p class="card-text">Genomics/Transcriptomics/Proteomics/Single Cell RNA-seq/Biological Network/Systems Biology</p>
      <ul class="network-icon" aria-hidden="true">
  
    
    
    
      
    
    
    
    
    
      
    
    <li>
      <a href="https://twitter.com/54zqz" target="_blank" rel="noopener">
        <i class="fab fa-twitter"></i>
      </a>
    </li>
  
    
    
    
      
    
    
    
    
    
      
    
    <li>
      <a href="https://github.com/zqzneptune" target="_blank" rel="noopener">
        <i class="fab fa-github"></i>
      </a>
    </li>
  
</ul>

    </div>
  </div>









  
  
  <div class="article-widget content-widget-hr">
    <h3>Related</h3>
    <ul>
      
      <li><a href="/post/2018-07-08-apms/">Awsome affinity purification mass spectrometry (AP-MS)</a></li>
      
      <li><a href="/publication/nbt.4024/">Global landscape of cell envelope protein complexes in Escherichia coli</a></li>
      
      <li><a href="/publication/j.cels.2017.10.010/">A Map of Human Mitochondrial Protein Interactions Linked to Neurodegeneration Reveals New Mechanisms of Redox Homeostasis and NF-κB Signaling</a></li>
      
      <li><a href="/publication/oncotarget.22413/">Renal oncocytoma characterized by the defective complex I of the respiratory chain boosts the synthesis of the ROS scavenger glutathione</a></li>
      
      <li><a href="/post/2014-09-23-rged/">Hello, RGED!</a></li>
      
    </ul>
  </div>
  



  </div>
</article>

      

    
    
    
      <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery.imagesloaded/4.1.4/imagesloaded.pkgd.min.js" integrity="sha256-lqvxZrPLtfffUl2G/e7szqSvPBILGbwmsGE1MKlOi0Q=" crossorigin="anonymous"></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery.isotope/3.0.6/isotope.pkgd.min.js" integrity="sha256-CBrpuqrMhXwcLLUd5tvQ4euBHCdh7wGlDfNz8vbu/iI=" crossorigin="anonymous"></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/fancybox/3.5.7/jquery.fancybox.min.js" integrity="sha256-yt2kYMy0w8AbtF89WXb2P1rfjcP/HTHLT7097U8Y5b8=" crossorigin="anonymous"></script>

      

      
        
        <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.15.10/highlight.min.js" integrity="sha256-1zu+3BnLYV9LdiY85uXMzii3bdrkelyp37e0ZyTAQh0=" crossorigin="anonymous"></script>
        
        <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.15.10/languages/r.min.js"></script>
        
      

    

    
    

    
    
    <script>const code_highlighting = true;</script>
    

    
    
    
    
    
    
    <script>
      const search_config = {"indexURI":"/index.json","minLength":1,"threshold":0.3};
      const i18n = {"no_results":"No results found","placeholder":"Search...","results":"results found"};
      const content_type = {
        'post': "Posts",
        'project': "Projects",
        'publication' : "Publications",
        'talk' : "Talks"
        };
    </script>
    

    
    

    
    
    <script id="search-hit-fuse-template" type="text/x-template">
      <div class="search-hit" id="summary-{{key}}">
      <div class="search-hit-content">
        <div class="search-hit-name">
          <a href="{{relpermalink}}">{{title}}</a>
          <div class="article-metadata search-hit-type">{{type}}</div>
          <p class="search-hit-description">{{snippet}}</p>
        </div>
      </div>
      </div>
    </script>
    

    
    
    <script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/3.2.1/fuse.min.js" integrity="sha256-VzgmKYmhsGNNN4Ph1kMW+BjoYJM2jV5i4IlFoeZA9XI=" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script>
    

    
    

    
    

    
    
    
    
    
    
    
    
    
      
    
    
    
    
    <script src="/js/academic.min.600a2e440c16093e23e26e90951c4d4b.js"></script>

    






  
  
  <div class="container">
    <footer class="site-footer">
  

  <p class="powered-by">
    ©2014-2020 Qingzhou Zhang (CC) BY-NC-SA 4.0. All thoughts and opinions here are my own. &middot; 

    Powered by the
    <a href="https://sourcethemes.com/academic/" target="_blank" rel="noopener">Academic theme</a> for
    <a href="https://gohugo.io" target="_blank" rel="noopener">Hugo</a>.

    
    <span class="float-right" aria-hidden="true">
      <a href="#" class="back-to-top">
        <span class="button_icon">
          <i class="fas fa-chevron-up fa-2x"></i>
        </span>
      </a>
    </span>
    
  </p>
</footer>

  </div>
  

  
<div id="modal" class="modal fade" role="dialog">
  <div class="modal-dialog">
    <div class="modal-content">
      <div class="modal-header">
        <h5 class="modal-title">Cite</h5>
        <button type="button" class="close" data-dismiss="modal" aria-label="Close">
          <span aria-hidden="true">&times;</span>
        </button>
      </div>
      <div class="modal-body">
        <pre><code class="tex hljs"></code></pre>
      </div>
      <div class="modal-footer">
        <a class="btn btn-outline-primary my-1 js-copy-cite" href="#" target="_blank">
          <i class="fas fa-copy"></i> Copy
        </a>
        <a class="btn btn-outline-primary my-1 js-download-cite" href="#" target="_blank">
          <i class="fas fa-download"></i> Download
        </a>
        <div id="modal-error"></div>
      </div>
    </div>
  </div>
</div>

</body>
</html>
